crowbar_framework/lib/crowbar/chef/upload.rb

```diff
@@ -0,0 +1,284 @@
+          cookbooks
+          data_bags
+          roles
+        end
+
+        def cookbooks
+          version_constraints_to_update = Hash.new
+          cookbooks_for_upload = Array.new
+
+          cookbook_loader.each do |cookbook_name, cookbook|
+            cookbooks_for_upload.push(cookbook)
+            version_constraints_to_update[cookbook_name] = cookbook.version
+          end
+
+          logger.info("Checking existing cookbooks...")
+          cookbooks_for_upload.reject! do |cookbook|
+            next unless api_coobbooks.include?(cookbook.name.to_s)
+            cookbook_exists?(cookbook)
+          end
+
+          unless cookbooks_for_upload.any?
+            logger.info("All cookbooks are already existing on the server")
+            return true
+          end
+
+          cookbooks_for_upload.each do |cookbook|
+            logger.info("Checking cookbook #{cookbook.name}...")
+            return false unless check_links(cookbook)
+            return false unless check_dependencies(cookbook)
+          end
+
+          logger.info("Uploading cookbooks... #{cookbooks_for_upload.map(&:name).join(", ")}")
+          uploader = ::Chef::CookbookUploader.new(
+            cookbooks_for_upload,
+            force: false,
+            concurrency: cookbooks_for_upload.count
+          )
+          return false unless uploader.upload_cookbooks
+
+          true
+        end
+
+        def data_bags
+          loader = ::Chef::Knife::Core::ObjectLoader.new(::Chef::DataBagItem, logger)
+          data_bags = loader.find_all_object_dirs(chef_data_bags_path) || []
+          data_bags.each do |data_bag|
+            data_bag_items = loader.find_all_objects(chef_data_bags_path.join(data_bag))
+            data_bag_item_paths = normalize_data_bag_item_paths(data_bag_items) || []
+            data_bag_item_paths.each do |data_bag_item_path|
+              # Workaround for a strange chef behavior
+              relative_path = chef_data_bags_path.relative_path_from(Pathname.new(Dir.pwd))
+              data_bag_item = loader.load_from(relative_path, data_bag, data_bag_item_path)
+              bag = ::Chef::DataBagItem.new
+              bag.data_bag(data_bag)
+              bag.raw_data = data_bag_item
+              logger.info("Uploading data_bag #{data_bag_item_path}")
+
+              return false unless bag.save
+            end
+          end
+          true
+        end
+
+        def roles
+          loader = ::Chef::Knife::Core::ObjectLoader.new(::Chef::Role, logger)
+          chef_roles_path.each_child do |file|
+            next unless file.file?
+            role = loader.load_from("roles", file)
+            logger.info("Uploading role #{file.basename}")
+
+            return false unless role.save
+          end
+          true
+        end
+
+        protected
+
+        def logger
+          return Rails.logger unless caller.grep(/rake/).present?
+          ::Logger.new(STDOUT)
+        end
+
+        def chef_path
```

Comment:
I liked it more like you had it before, `chef_path` is a bit misleading I think as it is only a directory of our creation. but I don't mind too much.